concrete/dense_directed_matrix.go

```diff
@@ -112,8 +115,16 @@ func (g *DirectedDenseGraph) HasEdgeFromTo(u, v graph.Node) bool {
 	return uid != vid && !isSame(g.mat.At(uid, vid), g.absent)
 }
 
-func (g *DirectedDenseGraph) Weight(e graph.Edge) float64 {
-	return g.mat.At(e.From().ID(), e.To().ID())
+func (g *DirectedDenseGraph) Weight(x, y graph.Node) (w float64, ok bool) {
+	xid := x.ID()
+	yid := y.ID()
+	if xid == yid {
+		return 0, true
```

Comment:
> Sane values for absent IMO are 0 (correlation graph), Inf (cost graph) and NaN (dubious but arguable). This gives the plausible behaviour for the cost graph case (a correlation graph would have self edges being 1 and I don't know what someone who is claiming an absent edge == NaN would want a self edge to be, it could be either of 0, 1 or Inf). Maybe there should be two parameters - self and absent. I don't like this very much, but maybe it's the best of a bad lot.

Having two parameters self and absent does not seem so bad if we consider correlation graphs, because now the code always returns 0. But calling Weight() with self edges is easily avoided, so we can postpone including g.self until it is clearer that it is really useful.

> Not sure. The cost of an edge to itself is alway 0 and the correlation is always 1. This is independent of whether the graph knows about it. We can add it though.

SGTM, there's no need to add it.